myst_nb/glue.py

```diff
@@ -0,0 +1,177 @@
+        later on.
+    variable: python object
+        A variable in Python for which you'd like to store its display value. This is
+        not quite the same as storing the object itself - the stored information is
+        what is *displayed* when you print or show the object in a Jupyter Notebook.
+    display: bool
+        Display the object you are gluing. This is helpful in sanity-checking the
+        state of the object at glue-time.
+    """
+    mimebundle, metadata = IPython.core.formatters.format_display_data(variable)
+    mime_prefix = "" if display else "application/papermill.record/"
+    metadata["scrapbook"] = dict(name=name, mime_prefix=mime_prefix)
+    ipy_display(
+        {mime_prefix + k: v for k, v in mimebundle.items()},
+        raw=True,
+        metadata=metadata,
+    )
+
+
+# Helper functions
+def find_glued_key(path_ntbk, key):
+    """Find an output mimebundle in a notebook based on a key."""
+    # Read in the notebook
+    if isinstance(path_ntbk, Path):
+        path_ntbk = str(path_ntbk)
+    ntbk = nbf.read(path_ntbk, nbf.NO_CONVERT)
+    outputs = []
+    for cell in ntbk.cells:
+        if cell.cell_type != "code":
+            continue
+
+        # If we have outputs, look for scrapbook metadata and reference the key
+        for output in cell["outputs"]:
+            meta = output.get("metadata", {})
+            if "scrapbook" in meta:
+                this_key = meta["scrapbook"]["name"].replace(mime_prefix, "")
+                if key == this_key:
+                    bundle = output["data"]
+                    bundle = {this_key: val for key, val in bundle.items()}
+                    outputs.append(bundle)
+    if len(outputs) == 0:
+        print(f"Warning: did not find key {this_key} in notebook {path_ntbk}")
+        return
+    if len(outputs) > 1:
+        print(
+            f"Warning: multiple variables found for key: {this_key}. Returning first value."
+        )
+    return outputs[0]
+
+
+def find_all_keys(ntbk, keys=None):
+    """Find all `glue` keys in a notebook and return a dictionary with key: outputs."""
+    if keys is None:
+        keys = {}
+
+    for cell in ntbk.cells:
+        if cell.cell_type != "code":
+            continue
+
+        for output in cell["outputs"]:
+            meta = output.get("metadata", {})
+            if "scrapbook" in meta:
+                this_key = meta["scrapbook"]["name"]
+                if this_key in keys:
+                    SPHINX_LOGGER.warn(
+                        f"Over-writing pre-existing glue key: {this_key}"
+                    )
+                keys[this_key] = output
+    return keys
+
+
+# Role and directive for pasting
+class Paste(SphinxDirective):
+    required_arguments = 1
+    final_argument_whitespace = True
+    has_content = False
+
+    option_spec = {"id": directives.unchanged}
+
+    def run(self):
+        return [PasteNode(self.arguments[0], "directive")]
```

Comment:
I think you need to add the source path / line number to the node here, for later use in reporting (have a look at other sphinx directives)